b2/src/main/java/org/jclouds/b2/binders/UploadFileBinder.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.b2.binders;
+
+import java.util.Map;
+
+import org.jclouds.http.HttpRequest;
+import org.jclouds.b2.domain.B2Object;
+import org.jclouds.b2.domain.UploadUrlResponse;
+import org.jclouds.b2.reference.B2Headers;
+import org.jclouds.rest.MapBinder;
+
+import com.google.common.base.Preconditions;
+import com.google.common.net.HttpHeaders;
+import com.google.common.net.PercentEscaper;
+
+public final class UploadFileBinder implements MapBinder {
+   public static final PercentEscaper escaper = new PercentEscaper("._-/~!$'()*;=:@", false);
+
+   @Override
+   public <R extends HttpRequest> R bindToRequest(R request, Map<String, Object> postParams) {
+      UploadUrlResponse uploadUrl = (UploadUrlResponse) postParams.get("uploadUrl");
+      B2Object b2Object = (B2Object) postParams.get("b2Object");
+      Preconditions.checkArgument(b2Object.fileId() == null, "fileId must be null, was %s", b2Object.fileId());
```

Comment:
One way we handle this (in Openstack providers) if to have 2 separate classes, one that is used to create an object, and one that is used when reading an object. Thus fields that can only be used for updating or reading cannot be specified incorrectly. AutoValue makes it easier to implement - you implement a regular autovalue for the "read" version of the class, and you implement an AutoValue with an AutoValue builder for the create/update (update could  be a separate class as well).

I also think that doing client-side validation checks is largely empty work and removing them helps to (sometimes completely) eliminate binders and mappers.

Thoughts?